### PR TITLE
Add support for checksum lists

### LINKED ISSUE #5130: Add support for hash sum files (tools like sha256sum)
Tools like `md5sum`, `shasum`, `sha1sum`, `sha256sum`, `sha3sum`, and `b2sum` output one line per input file:
- hash digest in hexadecimal
- space
- space or asterisk (indicates text or binary mode, respectively)
- file name
- newline

If the file was stdin, the file name is `-`.

For example, in binary mode, the output looks like this:
```
0000000000000000000000000000000000000000 *file0
0000000000000000000000000000000000000001 *file1
```

In text mode, the output looks like this:
```
0000000000000000000000000000000000000000  file0
0000000000000000000000000000000000000001  file1
```

`cksum` is similar, but it outputs the checksum in decimal, a space, the file size (with all trailing null bytes trimmed), a space, and the file name. If the file was stdin, the last two parts are omitted.

### PR DESCRIPTION
**Pull request by theonlypwner**
<dl>
<dt>EDIT:</dt>
<dd><q>I'm checking off each extension as I find the time to. I can't review everything at once, so I'm ticking off extensions/filenames as they're verified. Sorry for the mess…</q><br/>
<cite>— @Alhadis</cite></dd>
<dt>EDIT Ⅱ:</dt>
<dd><q>Never mind, this is <a href="https://github.com/github/linguist/issues/5756">@lildude's problem</a> now. 😁</q><br/><cite>— <i>ibid</i></cite></dd></dl>

## Description
Fixes #5130.

## Checklist
-	[x] **I am adding a new language.**
	-	[x] **The language is in-use by hundreds of users on GitHub.**
		- [x] Extensions:
			- [x] [`.sha256`](https://github.com/search?q=extension%3Asha256): ~789,426 results
			- [x] [`.crc`](https://github.com/search?q=extension%3Acrc) 12,509
			- [x] [`.crc32`](https://github.com/search?q=extension%3Acrc32) 5,262
			- [x] [`.digest`](https://github.com/search?q=extension%3Adigest) 3,071
			- [x] [`.md2`](https://github.com/search?q=extension%3Amd2) 41,384 (3D model files need to be excluded)
			- [x] [`.md4`](https://github.com/search?q=extension%3Amd4+-%22BEGIN+CERTIFICATE+REQUEST%22) 120
			- [x] [`.md5`](https://github.com/search?q=extension%3Amd5) 4,906,611
			- ~~[`.rmd160`](https://github.com/search?q=extension%3Armd160) 49~~
			- [x] [`.sha1`](https://github.com/search?q=extension%3Asha1) 2,782,935
			- [x] [`.sha2`](https://github.com/search?q=extension%3Asha2) 423
			- [x] [`.sha224`](https://github.com/search?q=extension%3Asha224+-%22BEGIN+CERTIFICATE+REQUEST%22) 224
			- [x] [`.sha256`](https://github.com/search?q=extension%3Asha256) 789,426
			- [x] [`.sha3`](https://github.com/search?q=extension%3Asha3) 28
			- [x] [`.sha384`](https://github.com/search?q=extension%3Asha384) 3,685
			- [x] [`.sha512`](https://github.com/search?q=extension%3Asha512) 154,632
		- [x] Filenames:
			- [x] [`SHA1SUMS`](https://github.com/search?q=filename%3ASHA1SUMS) 3,921
			- [x] [`SHA256SUMS`](https://github.com/search?q=filename%3ASHA256SUMS) 2,148
			- …others (according to the template, only extensions need to be declared)
	- [x] **I have included real-world samples for all extensions added in this PR:**
		- **Source:** Author (@theonlypwner)
		- **License:**
			> If it isn't already in the public domain, I agree to license my samples under the MIT License, so it may be included in this repo without any issue.
	- [x] **I have included or enabled a [syntax highlighting][fig-1] grammar.**
	- [x] **I have added or updated the heuristics.**

[fig-1]: https://github-lightshow.herokuapp.com/?utf8=%E2%9C%93&scope=from-url&grammar_format=auto&grammar_url=https%3A%2F%2Fgithub.com%2FAlhadis%2Flanguage-etc%2Fblob%2F2f8309a73fcffcd82ebf158e8c202d6a3ae31f80%2Fgrammars%2Fchecksums.cson&grammar_text=&code_source=from-url&code_url=https%3A%2F%2Fgithub.com%2Ftheonlypwner%2Flinguist%2Fblob%2Fchecksums%2Fsamples%2FChecksums%2FSHA256SUMS&code=

### CODE REVIEW

File: lib/linguist/languages.yml [hunk 1]
```diff
@@ -766,6 +766,32 @@ Charity:
...
   language_id: 56
+Checksums:
+  type: data
+  aliases:
+    - hash
+    - hashes
+    - sum
+    - sums
+  filenames: [
+    "sum",      "sums",
```
Comment (smola) [inline lib/linguist/languages.yml:777]: Names must be sorted alphabetically. Also, all entries use the bullet syntax for lists, instead of `[...]`.

**Review by smola — CHANGES_REQUESTED** [review]

File: lib/linguist/languages.yml (hunk 1 above)
Comment (theonlypwner) [reply lib/linguist/languages.yml]: I originally wanted to group names by hash type, but I've changed it to meet that requirement.

Comment (Alhadis) [reply lib/linguist/languages.yml]: @theonlypwner Trouble is, some of these filenames might not meet the usage requirements, which means they'll end up being removed (forcing you to regroup or reformat the remaining entries).

File: lib/linguist/languages.yml [hunk 2]
```diff
@@ -766,6 +766,122 @@ Charity:
...
+  - sha384
+  - sha384sum
+  - sha384sums
+  - sha3sum
+  - sha3sums
+  - sha512
+  - sha512sum
+  - sha512sums
+  - sum
+  - sums
```
Comment (Alhadis) [inline lib/linguist/languages.yml:868]: Filenames, like extensions, are tested case-insensitively (although I vaguely recall there being regression-tests that *did* match filenames case-sensitively).

In any case, the entries named after a hash-function or abbreviation are generally uppercased, so let's golf this list down to:

~~~suggestion
  filenames:
  - CRC
  - CRC32
  - CRC32SUM
  - CRC32SUMS
  - CRCSUM
  - CRCSUMS
  - MD2
  - MD2SUM
  - MD2SUMS
  - MD4
  - MD4SUM
  - MD4SUMS
  - MD5
  - MD5SUM
  - MD5SUMS
  - RMD160
  - RMD160SUM
  - RMD160SUMS
  - SHA1
  - SHA1SUM
  - SHA1SUMS
  - SHA2
  - SHA224
  - SHA224SUM
  - SHA224SUMS
  - SHA256
  - SHA256SUM
  - SHA256SUMS
  - SHA2SUM
  - SHA2SUMS
  - SHA3
  - SHA384
  - SHA384SUM
  - SHA384SUMS
  - SHA3SUM
  - SHA3SUMS
  - SHA512
  - SHA512SUM
  - SHA512SUMS
  - checksum
  - checksums
  - cksum
  - cksums
  - sum
  - sums
~~~

File: samples/Checksums/single_hash.crc [hunk 3]
```diff
@@ -0,0 +1 @@
+ffffffff
```
Comment (Alhadis) [inline samples/Checksums/single_hash.crc:1]: IMO, the `.crc` extension should be removed. Even in hexadecimal, something innocuous and random has a high risk of matching something unrelated:

~~~console
$ ls -1 github.com/random-designer/website-mockup/swatches/solid-colours/*
background.crc
foreground.crc
~~~

Here, "CRC" could stand for an ad-hoc abbreviation of _"Centred Rectangle Colour"_, containing the hex-code for an RGBA (or ARGB) colour:

~~~cson
# solid-black.crc
00000000
~~~

Not saying this is likely, but more that this is the sort of random junk that tends to frequent search-results for a very generic-looking extension. I'm basing that on precedent search efforts... you wouldn't believe the amount of weird and wonderful crap people upload to GitHub. 😉

It'd be different if `.crc` files had a common-enough syntax that we could target with a heuristic, e.g.,

~~~py
CRC32 (filename) = 00000000
~~~

… but I doubt that's commonplace enough (haven't checked).

**Review by Alhadis — CHANGES_REQUESTED** [review]
Have yet to verify the usage of each addition, but here're some preliminary thoughts.

File: samples/Checksums/single_hash.crc (hunk 3 above)
Comment (theonlypwner) [reply samples/Checksums/single_hash.crc:1]: I didn't see BSD-style checksums for `.crc`, but I saw [another format similar to INI files on the second page](https://github.com/search?p=2&q=extension%3Acrc+CRC32):
```
filename=FILE
size=1337
crc32=00000000
```

**Review by Alhadis — APPROVED** [review]
@lildude, I'm passing this ball to your court, as you're currently responsible for sussing out popularity metrics. Hopefully it's not too much of a pain. 😁

File: lib/linguist/languages.yml [hunk 4]
```diff
@@ -940,6 +940,75 @@ Charity:
...
+  - SHA384SUM
+  - SHA384SUMS
+  - SHA3SUM
+  - SHA3SUMS
+  - SHA512
+  - SHA512SUM
+  - SHA512SUMS
+  - checksum
+  - checksums
+  - cksum
```
Comment (lildude) [inline lib/linguist/languages.yml:993]: Not popular enough, though checksums.txt is:

```suggestion
  - checksums.txt
```

File: lib/linguist/languages.yml [hunk 5]
```diff
@@ -940,6 +940,75 @@ Charity:
...
+  - cksums
+  - sum
+  - sums
```
Comment (lildude) [inline lib/linguist/languages.yml:996]: Not popular enough

```suggestion
```

File: lib/linguist/languages.yml [hunk 6]
```diff
@@ -940,6 +940,75 @@ Charity:
...
+  extensions:
+  - ".crc32"
+  - ".md2"
+  - ".md4"
+  - ".md5"
+  - ".rmd160"
```
Comment (lildude) [inline lib/linguist/languages.yml:1002]: Not popular enough:

```suggestion
```

**Comment by theonlypwner** [comment]
I agree with the suggestions and have made them.

I noticed a lot of results for extension `.digest`, so please check it.

Also, I noticed some files in the format `name whitespace hash` instead of `hash whitespace name`, which I hope either won't cause issues for the syntax highlighter or can be fixed later.

**Comment by lildude** [comment]
> I noticed a lot of results for extension `.digest`, so please check it.

Yes, there are quite a few, however the format isn't consistent from a quick look at various results. I'm not sure if the grammar can support this at this stage and I'd really like to build the next release today.

We can always add it later once @Alhadis confirms the grammar can support these various formats.

Accordingly, I've removed the `.digest` and the sample for now.

**Comment by theonlypwner** [comment]
> We can always add it later

Yes, I agree. It's better to merge this first and open a new issue/PR for later changes.

**Comment by Alhadis** [comment]
> I'm not sure if the grammar can support [`.digest` files] at this stage and I'd really like to build the next release today.

Well, I can add support for patterns [like this](https://cs.github.com/geosolutions-it/ogc-testbed14-styles/blob/03d9b77d22953308eee88743f34ac4dc3e13ef0b/security/masterpw.digest?q=path%3A*.digest+NOT+nothack#L1)…

~~~
digest1:dwLYXWqDkzaJhlo1P6sQqpxgU5cBRRYM9Sna/buqhHIoAX6/np9fYp9Zbgb3Rc7I
~~~

… however, I'm unfamiliar with whatever program generates this kind of output, and I'd ideally like some authoritative reference on what formats are accepted in such lines (the example looks to be using base64, but I'd rather not go by guesswork if I can help it).

> Also, I noticed some files in the format `name whitespace hash` instead of `hash whitespace name`, which I hope either won't cause issues for the syntax highlighter or can be fixed later.

There's no foolproof way to tell `name hash` from a filename that contains spaces, followed by a word that happens to contain <var>n</var> alphanumeric characters. Conversely, a line that begins with precisely <var>n</var> characters is a little less ambiguous. Consider files that begin with arbitrary text (e.g., [comments](https://github.com/ssdi-drive/nuxeo-drive/blob/8cc64b77cc464ff87cf9f8959d0a6f7085627375/tools/checksums.txt)), or even files [like this](https://github.com/molgryn/facebook-clang-plugins/blob/18bf6770420d4d53ce5bbc1bfbee909f3662c9d9/clang/src/checksums.txt#L4) that aren't technically a checksum list, but are still highlighted non-intrusively.

**Pull request merged**